Proofs: locals and memory identity proofs

hjorthjort opened the pull request:
Two new prover tests:
- One checking that loading a local and storing it back leaves the state unchanged.
- One checking the same for memory load followed by store.

This also removes the memory helpers `#clearRange` and `[_ := _]` and replaces them by a single function, `#setRange`, which combines their behavior.

The trusted computing base is extended with a lemma that is fairly clear from the _intended_ semantics of how stores and loads from Wasm memory happens, but that we haven't proven holds for our implementations of the functions.

hjorthjort commented:
Seems the two failing CI-runs were due to timeouts. Maybe I got a really slow machine, but should we up the timeout anyway?

ehildenb commented on data.md:
```diff
@@ -492,51 +492,42 @@ Byte Map
+    rule #get(               M, KEY ) => 0 requires notBool KEY in_keys(M) [concrete]
+
+    rule #set( M, KEY, VAL ) => M [KEY <- undef] requires VAL ==Int 0  [concrete]
+    rule #set( M, KEY, VAL ) => M [KEY <- VAL  ] requires VAL  >Int 0  [concrete]
```
Please make these total and mutually exclusive. I'd rather the side-condition be `notBool VAL ==Int 0`, since it makes it obvious that these are total and mutually exclusive.

Of course, the semantics will only ever exercise the `VAL >Int 0` case, but it makes it easier to do reasoning in the Haskell backend when functions are total.

hjorthjort replied:
Done

ehildenb commented:
@hjorthjort let's run the proofs in their own stage with their own timeouts, since you are adding proofs as we go and we're bound to keep needing to increase the timeouts on the proofs.


I would rename `smoke` to `simple`, then run `conformance` after `simple`, and run `proofs` after `conformance`.

ehildenb commented:
I get to a final state that looks like this:

```
#set(
     #set(
          #set(
               #set(
                    R_DATA,
                    EA,
                    _modInt_(#get(R_DATA, EA) , 256))  # Use ensures to enforce R_DATA is 256?
               , (EA + 1)
               , _modInt_((#get(R_DATA, EA) / 256) , 256)
               )
          , ((EA + 1) + 1)
          , _modInt_(((#get(R_DATA, EA) / 256) / 256) , 256)
          )
     , (((EA + 1) + 1) + 1) , _modInt_((((#get(R_DATA, EA) / 256) / 256) / 256) , 256)
     )
```

Using the following lemmas

```
    rule #get(#set(MAP, IDX, VAL), IDX) => VAL
    rule #set(MAP, IDX, #get(MAP, IDX)) => MAP

    rule (X +Int (Y *Int N)) modInt N => X modInt N
      requires N >=Int 1
    rule (X *Int N) /Int N => X 
      requires N =/=Int 0
```

And the definitions:

```
    syntax Map ::= #setRange(Map, Int, Int, Int) [function]
 // -------------------------------------------------------
    rule #setRange(BM, IDX,   _, N) => BM
      requires N  ==Int 0
    rule #setRange(BM, IDX, VAL, N) => #setRange(#set(BM, IDX, VAL modInt 256), IDX +Int 1, VAL /Int 256, N -Int 1)
      requires N =/=Int 0

    syntax Int ::= #getRange ( Map , Int , Int ) [function]
 // -------------------------------------------------------
    rule #getRange(BM, START, WIDTH) => 0
      requires WIDTH  ==Int 0
    rule #getRange(BM, START, WIDTH) => #get(BM, START) +Int (#getRange(BM, START +Int 1, WIDTH -Int 1) *Int 256)
      requires WIDTH =/=Int 0

    syntax Int ::= #get ( Map , Int      ) [function, smtlib(mapGet)]
    syntax Map ::= #set ( Map , Int, Int ) [function, smtlib(mapSet)]
 // -----------------------------------------------------------------
    rule #get( M, KEY ) => M [KEY] requires         KEY in_keys(M) [concrete]
    rule #get( M, KEY ) => 0       requires notBool KEY in_keys(M) [concrete]

    rule #set( M, KEY, VAL ) => M [KEY <- undef] requires          VAL ==Int 0  [concrete]
    rule #set( M, KEY, VAL ) => M [KEY <- VAL  ] requires notBool (VAL ==Int 0) [concrete]
```

I'm not sure how I'll get the modInt around #get to go away. I'll have a deeper look tomorrow and see if I can think of a way to restructure the rules (maybe the enforcement of the maps containing values between 1 and 255 should be in #set and #get?)

ehildenb commented on data.md:
```diff
@@ -491,51 +491,42 @@ Byte Map
-    rule #range(BM:Map, START, WIDTH) => #lookup(BM, START) +Int (#range(BM, START +Int 1, WIDTH -Int 1) *Int 256)
-      requires WIDTH >Int 0 [concrete]
+    rule #getRange(BM, START, WIDTH) => #get(BM, START) +Int (#getRange(BM, START +Int 1, WIDTH -Int 1) *Int 256)
+      requires WIDTH >Int 0
```
Let's use `WIDTH >Int 0` and `notBool (WIDTH >Int 0)` for side conditions, which (1) ensures totalness and (2) ensures termination.

hjorthjort replied:
Done

ehildenb commented on data.md:
```diff
@@ -491,51 +491,42 @@ Byte Map
+    rule #setRange(BM, IDX,   _, N) => BM
+      requires N  ==Int 0
+    rule #setRange(BM, IDX, VAL, N) => #setRange(#set(BM, IDX, VAL modInt 256), IDX +Int 1, VAL /Int 256, N -Int 1)
+      requires N =/=Int 0
```
Let's use `N >Int 0` and `notBool (N >Int 0)` for side conditions, to (1) ensure completeness and (2) ensure termination.

hjorthjort replied:
Done

hjorthjort commented:
Hmm, something is off in the final state. The index in the `#get` calls doesn't increase, it's always `EA`. Can't figure out why, `#getRange` certainly increments the index.

hjorthjort commented:
Resolved it. I was using an incorrect lemma (!), saying that `(x + y*n) / n == x / n` (!). Fixed the lemmas,  added som other basic arithmetic `x + 0 == 0 + x = x`, and the proof goes through without any high level lemma on `[get|set]Range`.

hjorthjort commented:
@ehildenb Two design decisions here, let me know what you think. 
1. Introduce the `ByteMap` sort, which is just a wrapper around a regular map. By making this special sort, we can make sure that `#(get/set)[Range]` only operate on ByteMaps, and not regular maps. That way, we can more confidently add lemmas about these functions work by making use of the `ByteMap` invariant of only containing bytes. (Commit 868de9e)
2. Introduce the `#isWellFormedByteMap` property. The value that it adds is that it forces us to explicitly assume that the map is well-formed in our proffs, rather than tacitly assuming it throughout the proofs, the way we do in our spec. Down the road we could also try to prove using induction that the semantics can't cause a violation of this property.

ehildenb commented on data.md:
```diff
@@ -491,51 +491,49 @@ Byte Map
 
-`#clearRange(MAP, START, END)` removes all entries in the map from `START` (inclusive) to `END` (exclusive).
+`#get` looks up a key in a map, defaulting to 0 if the map does not contain the key.
+`#set` sets a key in a map, defaulting to 0 if the value is 0.
```
```suggestion
`#set` sets a key in a map, unsetting that key if the value is 0.
```

ehildenb commented on kwasm-lemmas.md:
````diff
@@ -31,6 +44,42 @@ In this case, it's simpler (and safe) to simply discard the `#chop`, instead of
+
+```k
+    syntax Bool ::= "#isWellFormedByteMap" "(" ByteMap ")" [function, smtlib(isWellFormedByteMap)]
+ // ----------------------------------------------------------------------------------------------
````
Do you want to add rules about what a well-formed `ByteMap` looks like? Something like:

```
rule #isByteMap(ByteMap <| .Map |>) => true
rule #isByteMap(ByteMap <| _ |-> V M |>) => #isByte(V) andBool #isByteMap(M)
rule #isByteMap(ByteMap <| M [ _ <- V ] |>) => #isByte(V) andBool #isByteMap(M)
```

Note that I also am suggesting the shorter name `#isByteMap` over `#isWellFormedByteMap`.

hjorthjort replied:
Good idea. I was mostly afraid this would trip up the prover (that it would start applying these rules and get stuck, rather than just take it at face value), but the proofs seem to go through still.
